drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py

```diff
@@ -0,0 +1,590 @@
+        try:
+            self._request_signals()
+            self._receive_signals()
+            T = self.getReceiverPose(rospy.Time.now(), self.receiver_array_frame,
+                                     self.locating_frame)
+            self.translation, self.rotation = T
+            self.ready = True
+        except Exception as e:
+            rospy.logerr(str(e))
+            raise e
+ 
+    def _request_signals(self):
+        '''
+        Request a set of digital signals from a serial port
+
+        serial_port - open instance of serial.Serial
+        data_request_code - char or string to sent to receiver board to signal a data request
+        tx_start_code - char or string that the receiver board will send  us to signal the
+            start of a transmission
+        '''
+        self.ser.flushInput()
+        try:
+            readin = None
+
+            # Request raw signal tx until start bit is received
+            while readin == None or ord(readin) != ord(self.tx_start_code):
+                self.ser.write(self.tx_request_code)
+                readin = self.ser.read(1)
+                if len(readin) < len(self.tx_request_code):  # serial read timed out
+                    raise IOError('Timed out waiting for serial response.')
+
+        except serial.SerialException as e:
+            rospy.logerr(str(e))
+            return None
+
+    def _receive_signals(self):
+        '''
+        Receives a set of 1D signals from a serial port and packs them into a numpy array
+
+        serial_port - port of type serial.Serial from which to read
+        signal_size - number of sacalar elements in each 1D signal
+        scalar_size - size of each scalar in bytes
+        signal_bias - value to be subtracted from each read scalar before packing into array
+
+        returns: 2D numpy array with <num_signals> rows and <signal_size> columns
+        '''
+        self.signals = np.full((self.num_receivers, self.signal_size), self.signal_bias, dtype=float)
+
+        for channel in range(self.num_receivers):
+            for i in range(self.signal_size):
+                while self.ser.inWaiting() < self.scalar_size:
+                    pass
+                self.signals[channel, i] = float(self.ser.read(self.scalar_size)) - self.signal_bias
+
+    def getReceiverPose(time, receiver_array_frame, locating_frame):
+        '''
+        Gets the pose of the receiver array frame w.r.t. the locating_frame
+        (usually /map or /world).
+
+        Returns a 3x1 translation and a 3x3 rotation matrix (numpy arrays)
+        '''
+        try:
+            tfl = tf2_ros.TransformListener(self.tf2_buf)
+            T = self.tf2_buf.lookup_transform(receiver_array_frame, locating_frame, time,
+                                              timeout=rospy.Duration(0.20))
+            q = T.transform.rotation
+            t = T.transform.translation
+            rot = transformations.quaternion_matrix([q.w, q.x, q.y, q.z])
```

Comment:
That's a good idea, I explicitly didn't do that or use other helpers to avoid the dependency on other mil pkgs.